lib/semantics/resolve-labels.cc

```diff
@@ -311,6 +320,156 @@ class ParseTreeAnalyzer {
+            std::get_if<std::optional<parser::GenericSpec>>(
+                &interfaceStmt.u)}) {
+      if (optionalGenericSpecPointer->has_value()) {
+        if (const auto *namePointer{
+                std::get_if<parser::Name>(&(*optionalGenericSpecPointer)->u)}) {
+          auto &optionalGenericSpec{
+              std::get<parser::Statement<parser::EndInterfaceStmt>>(
+                  interfaceBlock.t)
+                  .statement.v};
+          if (optionalGenericSpec.has_value()) {
+            if (const auto *otherPointer{
+                    std::get_if<parser::Name>(&optionalGenericSpec->u)}) {
+              if (namePointer->ToString() != otherPointer->ToString()) {
+                errorHandler_.Say(currentPosition_,
+                    parser::MessageFormattedText{
+                        "INTERFACE generic-name (%s) mismatch"_en_US,
+                        namePointer->ToString().c_str()});
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+  // C1402
+  void Post(const parser::Module &module) {
+    if (!PresentAndEq(
+            std::get<parser::Statement<parser::EndModuleStmt>>(module.t)
+                .statement.v,
+            std::get<parser::Statement<parser::ModuleStmt>>(module.t)
+                .statement.v)) {
+      errorHandler_.Say(currentPosition_,
+          parser::MessageFormattedText{"END MODULE name mismatch"_err_en_US});
+    }
+  }
+  // C1569
+  void Post(const parser::SeparateModuleSubprogram &separateModuleSubprogram) {
+    if (!PresentAndEq(std::get<parser::Statement<parser::EndMpSubprogramStmt>>(
+                          separateModuleSubprogram.t)
+                          .statement.v,
+            std::get<parser::Statement<parser::MpSubprogramStmt>>(
+                separateModuleSubprogram.t)
+                .statement.v)) {
+      errorHandler_.Say(currentPosition_,
+          parser::MessageFormattedText{
+              "END MODULE PROCEDURE name mismatch"_err_en_US});
+    }
+  }
+  // C1401
+  void Post(const parser::MainProgram &mainProgram) {
+    if (std::get<std::optional<parser::Statement<parser::ProgramStmt>>>(
+            mainProgram.t)
+            .has_value()) {
+      auto &programStmt{
+          std::get<std::optional<parser::Statement<parser::ProgramStmt>>>(
+              mainProgram.t)
+              ->statement};
+      auto &endProgramStmt{
+          std::get<parser::Statement<parser::EndProgramStmt>>(mainProgram.t)
+              .statement};
+      if (endProgramStmt.v.has_value()) {
+        if (programStmt.v.ToString() != endProgramStmt.v->ToString()) {
+          errorHandler_.Say(currentPosition_,
+              parser::MessageFormattedText{
+                  "END PROGRAM name mismatch"_err_en_US});
+        }
+      } else {
+        errorHandler_.Say(currentPosition_,
+            parser::MessageFormattedText{
+                "END PROGRAM must have a program-name"_err_en_US});
```

Comment:
The name is optional on END PROGRAM.

Reply:
I'll fix this. The constraint reads "The program-name shall not be included in the end-program-stmt unless the optional program-stmt is used. _If included_, it shall be identical to the program-name specified in the program-stmt." and I confused myself about what optional object in the first sentence was implicitly referred to by "if included." Thanks.